rfcs/0001-messages.md

```diff
@@ -0,0 +1,278 @@
+
+This will require deriving serde in several types defined in `frost.rs`. 
+Manual implementation of serialization/deserialization will be located at a new mod `src/frost/serialize.rs`.
+
+### Byte order
+
+Each byte chunk specified below is in little-endian order unless is specified otherwise.
+
+### Header
+
+The `Header` part of the message is 4 bytes total:
+
+Bytes | Field name | Data type
+------|------------|-----------
+1     | msg_type   | u8
+1     | version    | u8
+1     | sender     | u8
+1     | receiver   | u8
+
+### Primitive types
+
+`Payload`s use data types that we need to specify first. We have 2 primitive types inside the payload messages:
+
+#### `Scalar`
+
+`Scalar` is a an alias for `jubjub::Fr` and this is a `[u64; 4]` as documented in https://github.com/zkcrypto/jubjub/blob/main/src/fr.rs#L16
+
+#### `AffinePoint`
+
+Much of the math in FROST is done using `jubjub::ExtendedPoint`. This is a structure with 5 `jubjub::Fq`s as defined in https://github.com/zkcrypto/jubjub/blob/main/src/lib.rs#L128-L134
+
+Each `Fq` needed to form a `jubjub::ExtendedPoint` are `Scalar`s of `bls12_381` crate. Scalar here is `[u64; 4]` as documented in https://github.com/zkcrypto/bls12_381/blob/main/src/scalar.rs#L16
+
+For message exchange `jubjub::AffinePoint`s are a better choice as they are shorter in bytes, they are formed of 2 `jubjub::Fq` instead of 5: https://github.com/zkcrypto/jubjub/blob/main/src/lib.rs#L70-L73
+
+Conversion from one type to the other is trivial:
+
+https://docs.rs/jubjub/0.6.0/jubjub/struct.AffinePoint.html#impl-From%3CExtendedPoint%3E
+https://docs.rs/jubjub/0.6.0/jubjub/struct.ExtendedPoint.html#impl-From%3CAffinePoint%3E
+
+### Payload
+
+Payload part of the message is variable in size and depends on message type.
+
+#### `MsgDealerBroadcast`
+
+Bytes  | Field name  | Data type
+-------|-------------|-----------
+256    | secret_key  | Scalar
+512    | commitments | AffinePoint
+256+32 | group_public| (AffinePoint, [u8; 32])
+
+#### `MsgCommitments`
+
+Bytes | Field name | Data type
+------|------------|-----------
+512   | hiding     | AffinePoint
+512   | binding    | AffinePoint
+
+#### `MsgSigningPackage`
+
+Bytes                  | Field name     | Data type
+-----------------------|----------------|-----------
+1                      | participants   | u8
+(1+256+256)*partipants | commitments    | Vec<(u8, AffinePoint, AffinePoint)>
+8                      | message_length | u64
+message_length         | message        | [u8]
+
+#### `SignatureShare`
+
+Bytes | Field name | Data type
+------|------------|-----------
+256   | signature  | Scalar
+
+#### `MsgFinalSignature`
+
+Bytes | Field name | Data type
+------|------------|-----------
+64    | signature  | ([u8; 32], [u8; 32])
```

Comment:
Would it be easier to split this tuple into two named fields?
Or to specify a `Signature` sub-type with two named fields?

Reply:
That might be good, as a signature consists of a `Commitment` and a `Response` which can be their own named types.

Reply:
`Signature` is https://github.com/ZcashFoundation/redjubjub/blob/frost-messages/src/signature.rs#L17-L21

Unsure what the `s_bytes` and `r_bytes` mean. I created a type but will be good to add documentation to the fields: 

```rust
// The final signature is broadcasted by the aggegator 
// to any participant.
struct MsgFinalSignature {
    // Bytes needed to build the frost::Signature
    final_signature: FinalSignature,
}

// Final RedJubJub signature the aggergator has created.
struct FinalSignature {
    //
    r_bytes: [u8; 32],
    //
    s_bytes: [u8; 32],
}
```